### DISCUSSION TITLE
Warning: a promise was rejected with a non-error: [object Object]

### DISCUSSION THREAD

**Discussion by simplecommerce** [vazco/uniforms#1047] [Q&A] [answered]
Hi,

I started seeing this warning recently in my devtools console whenever I validate a form.
It seems to point to `onValidate`.

I was wondering if anyone is seeing this also or is it just me?

**Comment by radekmie** [comment]
From what I checked, it's most likely a Bluebird warning ([docs](http://bluebirdjs.com/docs/warning-explanations.html#warning-a-promise-was-rejected-with-a-non-error)) - can you confirm that? And I guess you're using JSON schema, right? If so, we'll update our docs on how to mitigate it by wrapping the `validator` result in an error.

**Comment by simplecommerce** [comment, marked as answer]
@radekmie Hi, I am using GraphQL schema, but I did the following to remove the warning, don't know if its the right way though but it 
seems to work.

I created a custom class

```
class AutoFormValidationError extends Error {
    constructor(details) {
        super("AutoForm Validation Error"); // (1)
        this.name = "AutoFormValidationError"; // (2)
        this.details = details;
    }    
}
```

And I return it in `onValidate` , my onValidate returns a promise.

```
resolve(new AutoFormValidationError(details));
```

**Reply by radekmie** [reply]
That's basically what I wanted to suggest. Another option is to use something like `Object.assign(new Error(), { details })`.

**Reply by simplecommerce** [reply]
> That's basically what I wanted to suggest. Another option is to use something like `Object.assign(new Error(), { details })`.

Ok cool! I just wasn't sure if using assign like that was the right way to do it or not. 
Thanks!!

**Comment by simplecommerce** [comment]
@radekmie I hope you could help again with this topic. I have a form which I use `onValidate` to do some async validation and I display errors inline.

This code works:

```

const onValidate = (model, error) => {
	let details = get(error, "details", []);

	if (!disableRequiredFieldsValidation){
		// append errors
		details = concat(details, schema.validateRequiredFields(model));
	}

	const errorObj = details.length ? { details: details } : error || null;

	// eslint-disable-next-line no-console
	if (!isEmpty(errorObj)) console.error(errorObj);

	if (!isUnmountingRef.current){
		return Promise.resolve(isFunction(onValidateAsProp) ? onValidateAsProp(model, errorObj) : errorObj);
	}
};

```

I have no issues and I see the errors appear. I wanted to refactor the code so it would look better, so I did the following:

```

const onValidate = async (model, error) => {
	if (!isUnmountingRef.current){
		let details = get(error, "details", []);

		if (!disableRequiredFieldsValidation){
			// append errors
			details = concat(details, schema.validateRequiredFields(model));
		}

		let errorObj = details.length ? { details: details } : error || null;
		errorObj = isFunction(onValidateAsProp) ? await onValidateAsProp(model, errorObj) : errorObj;

		// eslint-disable-next-line no-console
		if (!isEmpty(errorObj)) console.log(errorObj);

		return errorObj;
	}
};

```

The 2nd code, no longer displays errors inline and in both cases I keep getting the ` a promise was rejected with a non-error: [object Object]` warning in my console.

I don't see what I am doing wrong or different. I tried to do the same test on a sandbox and the sandbox works fine.
My app is using MeteorJS but the sandbox isn't, so I am thinking maybe it has something to do with that, but it would be weird though.

Any help is appreciated, thanks!

**Reply by simplecommerce** [reply]
Hmm, it is weird, this works:

```

const onValidate = (model, error) => {
	return Promise.resolve(error);
};

```

But this doesn't:

```

const onValidate = async (model, error) => {
	return error;
};

```

I don't get the logic on why it wouldn't work at all.

**Reply by wadamek65** [reply]
@simplecommerce Could it be that the `Promise.resolve` that you use in the first example is perhaps a different `Promise` object than the one used by the `async` function? Maybe there is some polyfilling or prototype chain overriding going on in the background?

**Reply by simplecommerce** [reply]
@wadamek65 I honestly have no idea, in my last comment above your post, I did a simple return of the error so it should make no difference in both cases right, they both should work? 

I am trying to reproduce it in another MeteorJS project which does not have all the stuff I have in my current project to see if I can still get the same issue.

**Reply by wadamek65** [reply]
@simplecommerce It is hard for me to say, perhaps the transpilation down from `async` function syntax to an actual `Promise` is somehow using a different `Promise` object which causes these errors? A simple debug like could tell you if this is indeed the problem: `console.log(onValidate() instanceof Promise)` - if I'm not mistaken this should return `true` if the global `Promise` is the same in both cases.

**Reply by simplecommerce** [reply]
@wadamek65 i think you might be right, I tried on a project from scratch and I am not able to reproduce the issue at all in all cases, they all work, so it must be something else conflicting, I will keep debugging, and thanks for your input I appreciate it!!

**Reply by simplecommerce** [reply]
@wadamek65 you are right, the simple console log shows that the `Promise.resolve` is true to be an instance of Promise, but the 2nd one, on the server side it says true but on the client side it returns false.

**Reply by simplecommerce** [reply]
@wadamek65 I am having no luck finding the cause of the async function not returning true as an instance of Promise.
In my test reproduction repo, I installed every package I have in my actual project and I still get true in my test repo but false in my actual project. 

I tried to look in every file to see if anything would actually override Promise, but I don't see anything.

Any other ideas? If it is not package related?

**Reply by wadamek65** [reply]
@simplecommerce Are the build steps the same? Because I would expect it's rather the build system causing this rather than the packages themselves. 

Unfortunately, I do not have any other ideas for you. I'm also not 100% sure that what I mentioned is the root of your problem but I think it's highly likely. I could help if you managed to provide me with a minimal reproduction repo, otherwise, maybe @radekmie will have an idea once he comes back to this.

**Reply by simplecommerce** [reply]
@wadamek65 You were right, I finally was able to pinpoint what was causing it to return false. It was a library I was including in an HTML file.
Removing it, made the console.log return true and the async code returns the errors correctly now.

Thanks for your help!!

**Reply by wadamek65** [reply]
@simplecommerce No problem, I'm glad your problem is resolved.

**Reply by radekmie** [reply]
The message itself comes from Bluebird. I know, because I actually [mentioned it in the docs](https://uniforms.tools/docs/api-bridges/#note-on-bluebird), right after this discussion was solved eight months ago :sweat_smile: In any case, glad you've solved it!

_(I'll note it down that we should extend the docs on this topic even further.)_
